// sencha-tern/plugin/ext/lib/cmdAutoDependencies.js
module.exports = {
    plugins:            { aliasPrefix: 'plugin.', typeProperty: ['ptype', 'type'] },
    features:           { aliasPrefix: 'feature.', typeProperty: ['ftype'] },
    items:              { aliasPrefix: 'widget.', typeProperty: ['xtype'], defaultTypeProperty: 'defaultType' },
    componentLayout:    { aliasPrefix: 'layout.', defaultType: 'auto' },
    layout:             { aliasPrefix: 'layout.', defaultType: 'auto' },
    triggers:           { aliasPrefix: 'trigger.', isKeyedObject: true },
    overflowHandler:    { aliasPrefix: 'box.overflow.' },
    dockedItems:        { aliasPrefix: 'widget.', typeProperty: ['xtype'], defaultType: 'toolbar' },
    tbar:               { aliasPrefix: 'widget.', typeProperty: ['xtype'], defaultType: 'toolbar' },
    bbar:               { aliasPrefix: 'widget.', typeProperty: ['xtype'], defaultType: 'toolbar' },
    fbar:               { aliasPrefix: 'widget.', typeProperty: ['xtype'], defaultType: 'toolbar' },
    lbar:               { aliasPrefix: 'widget.', typeProperty: ['xtype'], defaultType: 'toolbar' },
    rbar:               { aliasPrefix: 'widget.', typeProperty: ['xtype'], defaultType: 'toolbar' },
    navigationModel:    { aliasPrefix: 'view.navigation.' },
    selectionModel:     { aliasPrefix: 'selection.' },
    proxy:              { aliasPrefix: "proxy.", defaultPropertyName : "defaultProxyType" },
    reader:             { aliasPrefix: "reader.", defaultPropertyName : "defaultReaderType"},
    writer:             { aliasPrefix: "writer.", defaultPropertyName : "defaultWriterType"},
    viewModel:          { aliasPrefix: 'viewmodel.' },
    controller:         { aliasPrefix: "controller." },
    store:              { aliasPrefix: "store.", defaultType: 'store' },
    header:             { aliasPrefix: "widget.", defaultType: "header" },
    buttons:            { aliasPrefix: "widget.", defaultType: "button" },
    axes:               { aliasPrefix: "axis." },
    interactions:       { aliasPrefix: "interaction." },
    series:             { aliasPrefix: 'series.' },
    identifier:         { aliasPrefix: 'data.identifier.' },
    fields:             { aliasPrefix: 'data.field.', defaultType: 'auto' },

    // mvc
    model:              { aliasPrefix: "model.", mvc: true },
    controllers:        { aliasPrefix: "controller.", mvc: true, blame: "all" },
    stores:             { aliasPrefix: "store.", mvc: true, isKeyedObject: true, defaultType: 'store' },
    models:             { aliasPrefix: "model.", mvc : true, blame: "all" },
    views:              { aliasPrefix: "view.", mvc: true, blame: "all" },
    profiles:           { aliasPrefix: "profile.", mvc: true, blame: "all" },
    autoCreateViewport: { aliasPrefix: "view.", mvc: true, requires: ["Ext.plugin.Viewport"] },
    links:              { aliasPrefix: "model.", mvc: true, isKeyedObject: true },

    // directRefs
    tabBar:             { directRef: 'Ext.tab.Bar' },
    enableOverflow:     { directRef: 'Ext.layout.container.boxOverflow.Menu' },
    session:            { directRef: 'Ext.data.Session' }
};
