// sencha-tern/plugin/ext/test/integration/parse_jsduck.js
"use strict";

var fs = require('fs');

/**
 * Divides members into configs, instanceProperties, staticProperties, and events
 * @param {Object} def A jsduck json definition
 * @returns {Object}
 */
module.exports = function parseJSDuck(file) {
    var def = JSON.parse(fs.readFileSync(file));

    var result = {
        name: def.name,
        private: def.private,
        protected: def.protected,
        singleton: def.singleton,
        deprecated: def.deprecated,
        configs: [],
        instanceProperties: [],
        staticProperties: [],
        events: []
    };

    def.members.forEach(function(member) {
        if (!member.private && !member.protected && !member.deprecated && !member.hide && !member.removed && member.name.indexOf('aria') !== 0) {
            switch (member.tagname) {
                case 'cfg':
                    result.configs.push(member.name);
                    break;
                case 'event':
                    result.events.push(member.name);
                    break;
                case 'property':
                case 'method':
                    if (member.static) {
                        result.staticProperties.push(member.name);
                    } else if (member.name !== 'constructor') {
                        result.instanceProperties.push(member.name);
                    }
            }
        }
    });

    return result;
};
